source/common/common/posix/thread_impl.cc

```diff
@@ -24,26 +26,89 @@ int64_t getCurrentThreadId() {
 
 } // namespace
 
-ThreadImplPosix::ThreadImplPosix(std::function<void()> thread_routine)
-    : thread_routine_(std::move(thread_routine)) {
-  RELEASE_ASSERT(Logger::Registry::initialized(), "");
-  const int rc = pthread_create(
-      &thread_handle_, nullptr,
-      [](void* arg) -> void* {
-        static_cast<ThreadImplPosix*>(arg)->thread_routine_();
-        return nullptr;
-      },
-      this);
-  RELEASE_ASSERT(rc == 0, "");
-}
+// See https://www.man7.org/linux/man-pages/man3/pthread_setname_np.3.html.
+// The maximum thread name is 16 bytes including the terminating nul byte,
+// so we need to truncate the string_view to 15 bytes.
+#define PTHREAD_MAX_LEN_INCLUDING_NULL_BYTE 16
+
+/**
+ * Wrapper for a pthread thread. We don't use std::thread because it eats exceptions and leads to
+ * unusable stack traces.
+ */
+class ThreadImplPosix : public Thread {
+public:
+  ThreadImplPosix(std::function<void()> thread_routine, OptionsOptConstRef options)
+      : thread_routine_(std::move(thread_routine)) {
+    if (options) {
+      name_ = (std::string(options->name_.substr(0, PTHREAD_MAX_LEN_INCLUDING_NULL_BYTE - 1)));
+    }
+    RELEASE_ASSERT(Logger::Registry::initialized(), "");
+    const int rc = pthread_create(
+        &thread_handle_, nullptr,
+        [](void* arg) -> void* {
+          static_cast<ThreadImplPosix*>(arg)->thread_routine_();
+          return nullptr;
+        },
+        this);
+    RELEASE_ASSERT(rc == 0, "");
+
+#ifdef __linux__
+    // If the name was not specified, get it from the OS. If the name was
+    // specified, write it into the thread, and assert that the OS sees it the
+    // same way.
+    if (name_.empty()) {
+      getNameFromOS(name_);
+    } else {
+      const int set_name_rc = pthread_setname_np(thread_handle_, name_.c_str());
+      if (set_name_rc != 0) {
+        ENVOY_LOG_MISC(trace, "Error {} setting name `{}': {}", set_name_rc, name_,
+                       strerror(set_name_rc));
```

Comment:
`strerror` usually takes errno as input. I don't think the rc from this function will give anything meaningful from `strerror` unless you see it documented as such somewhere.

Reply:
good question. The doc in https://www.man7.org/linux/man-pages/man3/pthread_setname_np.3.html doesn't specifically say that strerror can be used to translate these error messages, but the example does seem to comingle errno, and it did seem to work (an error printed during development looked right). Happy to back this out if you think this doesn't make sense.

Reply:
If it works, it's probably fine. 

Also, strerror is listed as not being threadsafe in the docs, but it's used in many places in the codebase. I think it seems like more trouble than it's worth to figure out the thread safety, and figure out if set_name_rc is valid to pass to it. I'd rather you just log the number. But I don't feel strongly about it.

Reply:
dropped strerror() no need to risk weird behavior when you can always look up the code later if needed.